lib/compress/zstd_lazy.c

```diff
@@ -503,10 +528,23 @@ size_t ZSTD_HcFindBestMatch_generic (
     U32 nbAttempts = 1U << cParams->searchLog;
     size_t ml=4-1;
 
+    const ZSTD_matchState_t* const dms = ms->dictMatchState;
+    const U32 ddsHashLog = dictMode == ZSTD_dedicatedDictSearch
+                         ? dms->cParams.hashLog - ZSTD_LAZY_DDSS_BUCKET_LOG : 0;
+    const size_t ddsIdx = dictMode == ZSTD_dedicatedDictSearch
+                        ? ZSTD_hashPtr(ip, ddsHashLog, mls) << ZSTD_LAZY_DDSS_BUCKET_LOG : 0;
+
+    U32 matchIndex;
+
+    if (dictMode == ZSTD_dedicatedDictSearch) {
+        const U32* entry = &dms->hashTable[ddsIdx];
+        PREFETCH_L1(entry);
```

Comment:
I made this super crude change (which may access out of bounds memory, but whatever). I saw a speed boost of ~10%. That confirms that we aren't prefetching early enough.


```suggestion
        const size_t ddsIdx1 = dictMode == ZSTD_dedicatedDictSearch
                            ? ZSTD_hashPtr(ip+1, ddsHashLog, mls) << ZSTD_LAZY_DDSS_BUCKET_LOG : 0;
        const U32* entry = &dms->hashTable[ddsIdx];
        const U32* entry1 = &dms->hashTable[ddsIdx1];
        PREFETCH_L1(entry);
        PREFETCH_L1(entry1);
```

Reply:
That likely also means that we could also prefetch the match pointers earlier too.